python/pyarrow/tests/test_io.py

```diff
@@ -105,6 +105,141 @@ def test_python_file_read():
     f.close()
 
 
+def test_python_file_readall():
+    data = b'some sample data'
+
+    buf = BytesIO(data)
+    with pa.PythonFile(buf, mode='r') as f:
+        assert f.readall() == data
+
+
+def test_python_file_readinto():
+    length = 10
+    data = b'some sample data longer than 10'
+    array = np.empty(length, dtype=np.uint8)
+    dst_buf = memoryview(array)
+    src_buf = BytesIO(data)
+
+    with pa.PythonFile(src_buf, mode='r') as f:
+        assert f.readinto(dst_buf) == 10
+
+        assert array[:length].tobytes() == data[:length]
+        assert len(array) == length
+
+
+def test_python_file_readinto_different_buffertype():
+    length = 5
+    data = b'some sample data longer than 10'
+    array = np.empty(length, dtype=np.uint16)
+    dst_buf = memoryview(array)
+    src_buf = BytesIO(data)
+    byte_length = array.nbytes
+
+    with pa.PythonFile(src_buf, mode='r') as f:
+        assert f.readinto(dst_buf) == byte_length
+
+        assert array[:byte_length].tobytes() == data[:byte_length]
+        assert len(array) == length
+
+
+def test_python_file_readline():
+    data = b'''line1
+    line2
+    line3
+    '''
+
+    buf = BytesIO(data)
+    buf2 = BytesIO(data)
+    with pa.PythonFile(buf, mode='r') as f:
+        for index in range(4):
+            assert f.readline() == buf2.readline()
+
+
+def test_python_file_readline_no_nl_at_eof():
+    data = b'''line1
+    line2
+    line3'''
+
+    buf = BytesIO(data)
+    buf2 = BytesIO(data)
+    with pa.PythonFile(buf, mode='r') as f:
+        for index in range(4):
+            assert f.readline() == buf2.readline()
+
+
+def test_python_file_readlines():
+    data = b'''line1
+    line2
+    line3
+    '''
+
+    buf = BytesIO(data)
+    buf2 = BytesIO(data)
+    with pa.PythonFile(buf, mode='r') as f:
+        assert f.readlines() == buf2.readlines()
+
+
+def test_python_file_readlines_with_hint():
+    data = b'''line1
+    line2
+    line3
+    '''
+
+    buf = BytesIO(data)
+    buf2 = BytesIO(data)
+    with pa.PythonFile(buf, mode='r') as f:
+        assert f.readlines(8) == buf2.readlines(8)
+
+
+def test_python_file_correct_abc():
+    with pa.PythonFile(BytesIO(b''), mode='r') as f:
+        assert isinstance(f, RawIOBase)
```

Comment:
It's dubious that this should be `RawIOBase`, since a `BytesIO` is buffered. OTOH this could be considered a best-effort registration.

Reply:
Is it specified somewhere? There is `BufferedIOBase`, but the doc strings on `(Raw)IOBase` don't seem to prescribe implementation details.

Reply:
Ok, I see what you mean. What do you think is the best course here: keep `RawIOBase` as ABC and test with, say, `FileIO` or switch to `IOBase` as ABC, since `PythonFile` can wrap any file-like object.

Reply:
We should choose between either `RawIOBase` and `BufferedIOBase`. I think it should depend mainly on the "short reads" question.

Reply:
At the system level, short reads can occur when reading from any file. When doing disk I/O, they are rather rare (but can occur if e.g. the system call is interrupted by a signal). They are quite common from a socket or any other character device.

However, our `FileRead` routine loops until the entire read is satisfied, so our IO classes should satisfy as buffered IO.

Reply:
I changed it to `BufferedIOBase`.